### ISSUE TITLE
Install on serverless pod

### ISSUE THREAD

**Issue by johnx102** [jeanjerome/EchoInStone#12]
Hi, did you think we can install this on serverless pod? Or we need a pod with constant ssh acces for do it working. I want replace whisperx serverless pod with you’re solution

**Comment by johnx102** [comment]
thanks I will try... thanks for you quick feedback

**Comment by johnx102** [comment]
I finally successfully lauch server less pod without issue , with serverless_main.py, which format for request is good for submit a file ?

**Comment by jeanjerome** [comment]
@johnx102 Big question :) 

Initially, I hadn't planned to support other deployment modes for this project, but I understand that serverless deployment can be interesting for many users. That’s why I accept your contribution and I've refactored the app to make it easier to support this new use case. You can find these changes in the `feature/serverless-integration` branch.

Could you please update your modifications and integrate them into this branch? Of course, if you have suggestions or see ways to improve the implementation, feel free to propose them—I'm open to your ideas!

Thanks a lot for your contribution!

**Comment by johnx102** [comment]
Thanks I try, look like good , when pod up I see the test process running , but after block with process running ....
look like running loop 2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | Testing serverless handler...\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | 2025-06-01 15:36:48,768 - INFO - EchoInStone.processing.pyannote_diarizer - Diarization successful for file: /app/test_results/Conversation_en_français_présentation_dialogue_n_1.wav\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta |   std = sequences.std(dim=-1, correction=1)\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | /usr/local/lib/python3.11/site-packages/pyannote/audio/models/blocks/pooling.py:104: UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at /pytorch/aten/src/ATen/native/ReduceOps.cpp:1831.)\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta |   warnings.warn(\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | \n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | See https://github.com/pyannote/pyannote-audio/issues/1370 for more details.\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta |    >>> torch.backends.cudnn.allow_tf32 = True\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta |    >>> torch.backends.cuda.matmul.allow_tf32 = True\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta |    >>> import torch\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | It can be re-enabled by calling\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | /usr/local/lib/python3.11/site-packages/pyannote/audio/utils/reproducibility.py:74: ReproducibilityWarning: TensorFloat-32 (TF32) has been disabled as it might lead to reproducibility issues and lower accuracy.\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | 2025-06-01 15:36:47,611 - INFO - EchoInStone.processing.whisper_audio_transcriber - Successfully transcribed: /app/test_results/Conversation_en_français_présentation_dialogue_n_1.wav\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | Due to a bug fix in https://github.com/huggingface/transformers/pull/28687 transcription using a multilingual Whisper will default to language detection followed by transcription instead of translation to English.This might be a breaking change for your use case. If you want to instead always translate your audio to English, make sure to pass `language='en'`.\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta |   warnings.warn(\n
2025-06-01  17:36:49.861 | info | kns1fkt0j46yta | /usr/local/lib/python3.11/site-packages/transformers/models/whisper/generation_whisper.py:573: FutureWarning: The input name `inputs` is deprecated. Please make sure to use `input_features` instead.\n
2025-06-01  17:36:49.860 | info | kns1fkt0j46yta | 2025-06-01 15:36:43,801 - INFO - EchoInStone.capture.youtube_downloader - Audio downloaded and converted to /app/test_results/Conversation_en_français_présentation_dialogue_n_1.wav\n
2025-06-01  17:36:49.860 | info | kns1fkt0j46yta | 2025-06-01 15:36:42,538 - INFO - EchoInStone.processing.pyannote_diarizer - Diarization pipeline loaded and set to use cuda.\n
2025-06-01  17:36:49.860 | info | kns1fkt0j46yta | 2025-06-01 15:36:41,658 - INFO - EchoInStone.processing.whisper_audio_transcriber - Transcription model and pipeline loaded successfully.\n
2025-06-01  17:36:49.860 | info | kns1fkt0j46yta | Device set to use cuda:0\n
2025-06-01  17:36:49.860 | info | kns1fkt0j46yta | 2025-06-01 15:36:39,778 - INFO - EchoInStone.processing.whisper_audio_transcriber - Using device: cuda:0 with dtype: torch.float16\n
2025-06-01  17:36:49.860 | info | kns1fkt0j46yta | 2025-06-01 15:36:39,778 - INFO - EchoInStone.app - Starting transcription process...\n
2025-06-01  17:36:49.860 | info | kns1fkt0j46yta | 2025-06-01 15:36:39,778 - INFO - root - Logging level set to: INFO\n
2025-06-01  17:36:49.860 | info | kns1fkt0j46yta | 2025-06-01 15:36:39,778 - INFO - root - Handlers configured: ['StreamHandler', 'RotatingFileHandler']\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta | }\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |   "message": "Successfully processed audio with 4 segments"\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |   "output_file": "speaker_transcriptions.json",\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |   ],\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |     ]\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |       " Bonjour, je m'appelle Barnab\u00e9, je suis belge et je viens de Bruxelles ? Alors on va peut-\u00eatre commencer par commander quelque chose \u00e0 manger."\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |       42.68,\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |       32.0,\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |       "SPEAKER_01",\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |     [\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |     ],\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |       " Bonjour, je m'appelle Cl\u00e9mence Monin et je viens de Suisse."\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |       31.0,\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |       26.0,\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |       "SPEAKER_02",\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |     [\n
2025-06-01  17:34:31.886 | info | kns1fkt0j46yta |     ],\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |       " Oui, alors je m'appelle Anna, je suis isra\u00e9lienne et je viens de Tel Aviv."\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |       26.0,\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |       19.0,\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |       "SPEAKER_03",\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |     [\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |     ],\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |       " Bonjour, bienvenue \u00e0 Montpellier, je m'appelle Philippe, je suis pr\u00e9sident de l'association J'arrive depuis trois ans. Alors merci d'\u00eatre venu. Alors pour commencer, on va se pr\u00e9senter chacun chacun notre tour. Est-ce qu'on peut commencer par toi, par exemple ?"\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |       19.0,\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |       1.0,\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |       "SPEAKER_00",\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |     [\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |   "transcriptions": [\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta |   "status": "success",\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta | {\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta | 2025-06-01 15:34:30,732 - INFO - EchoInStone.app - Speaker SPEAKER_01 (32.00s to 42.68s):  Bonjour, je m'appelle Barnabé, je suis belge et je viens de Bruxelles ? Alors on va peut-être commencer par commander quelque chose à manger.\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta | 2025-06-01 15:34:30,732 - INFO - EchoInStone.app - Speaker SPEAKER_02 (26.00s to 31.00s):  Bonjour, je m'appelle Clémence Monin et je viens de Suisse.\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta | 2025-06-01 15:34:30,732 - INFO - EchoInStone.app - Speaker SPEAKER_03 (19.00s to 26.00s):  Oui, alors je m'appelle Anna, je suis israélienne et je viens de Tel Aviv.\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta | 2025-06-01 15:34:30,732 - INFO - EchoInStone.app - Speaker SPEAKER_00 (1.00s to 19.00s):  Bonjour, bienvenue à Montpellier, je m'appelle Philippe, je suis président de l'association J'arrive depuis trois ans. Alors merci d'être venu. Alors pour commencer, on va se présenter chacun chacun notre tour. Est-ce qu'on peut commencer par toi, par exemple ?\n
2025-06-01  17:34:31.885 | info | kns1fkt0j46yta | 2025-06-01 15:34:30,732 - INFO - EchoInStone.app - Transcriptions have been saved to speaker_transcriptions.json\n
2025-06-01  17:34:31.884 | info | kns1fkt0j46yta | 2025-06-01 15:34:30,732 - INFO - EchoInStone.utils.data_saver - Data saved to test_results/

**Comment by jeanjerome** [comment]
Thanks for your detailed feedback and the logs—it helps a lot!

From what I see, the handler itself (kubernetes_handler) processes the audio and prints the result, but then the process does not exit as expected. This can sometimes happen if a background thread or process is still running—often due to dependencies (like deep learning libraries or audio tools) that start worker threads in the background.

As a quick fix for the local test and Kubernetes pod batch mode, you can add sys.exit(0) at the end of the script, just after printing the results. This will ensure the process exits cleanly once processing is complete, and the pod will terminate as expected.

```python
if __name__ == "__main__":
    # For testing purposes
    test_event = {
        'echo_input': 'https://www.youtube.com/watch?v=plZRCMx_Jd8',
        'output_dir': 'test_results'
    }

    print("Testing serverless handler...")
    result = kubernetes_handler(test_event)
    print(json.dumps(result, indent=2))
    sys.exit(0)  # Ensure the process exits after processing
```

Let me know if this solves the issue for you, or if you notice any lingering processes or threads. If the problem persists, please share any additional details about your environment, and we can investigate further!

Thanks again for your contribution and feedback!

**Comment by johnx102** [comment]
Ok i will test and tell you, for info I test it on serverless on run pod.
i had to modify little bit Dockerfile for pod start correctly

**Comment by johnx102** [comment]
for info I use this Dockerfile for do it working # ============================
# ✅ Dockerfile EchoInStone (RunPod Ready)
# ============================

FROM python:3.11-slim

ENV DEBIAN_FRONTEND=noninteractive

# Installer dépendances système
RUN apt-get update && apt-get install -y --no-install-recommends \
    ffmpeg git curl build-essential \
    && apt-get clean && rm -rf /var/lib/apt/lists/*

# Copier le projet
WORKDIR /app
COPY . /app

# Installer les dépendances Python
RUN pip install --upgrade pip && \
    pip install -r requirements.txt

# Définir la commande de démarrage
CMD ["python3", "main.py"]

**Comment by jeanjerome** [comment]
@johnx102 Could you please describe the issue you're experiencing? I don't see anything unusual in the logs. Thank you.

**Comment by johnx102** [comment]
here my working handler.py """
RunPod Serverless handler for EchoInStone audio processing.
"""

import json
import logging
import os
import sys
import runpod

# Add the parent directory to the path to import EchoInStone
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from EchoInStone.utils import configure_logging
from EchoInStone.app import EchoInStoneApp

# Configure logging once at module level
log_level = os.getenv('LOG_LEVEL', 'INFO').upper()
configure_logging(getattr(logging, log_level))
logger = logging.getLogger(__name__)

def handler(job):
    """
    RunPod serverless handler function.
    
    Args:
        job: RunPod job object containing:
            - job["input"]: Dictionary with request parameters
            - job["id"]: Unique job identifier
    
    Returns:
        Dict containing the processing result
    """
    job_id = job.get('id', 'unknown')
    logger.info(f"EchoInStone | Starting job {job_id}")
    
    try:
        # Extract parameters from job input
        job_input = job.get("input", {})
        echo_input = job_input.get('echo_input')
        output_dir = job_input.get('output_dir', '/tmp/results')
        
        # Speaker diarization options
        num_speakers = job_input.get('num_speakers')  # Force exact number
        min_speakers = job_input.get('min_speakers')  # Minimum speakers
        max_speakers = job_input.get('max_speakers', 2)  # Maximum speakers (default: 2)
        
        # Validate required parameters
        if not echo_input:
            error_msg = 'Missing required parameter: echo_input'
            logger.error(f"Job {job_id} | {error_msg}")
            return {
                'status': 'error',
                'error': error_msg,
                'job_id': job_id
            }
        
        logger.info(f"Job {job_id} | Processing audio from: {echo_input}")
        logger.info(f"Job {job_id} | Output directory: {output_dir}")
        logger.info(f"Job {job_id} | Speaker constraints: max={max_speakers}, min={min_speakers}, exact={num_speakers}")
        
        # Ensure output directory exists
        os.makedirs(output_dir, exist_ok=True)
        
        # Prepare diarization options
        diarization_options = {}
        if num_speakers:
            diarization_options['num_speakers'] = num_speakers
        elif min_speakers or max_speakers:
            if min_speakers:
                diarization_options['min_speakers'] = min_speakers
            if max_speakers:
                diarization_options['max_speakers'] = max_speakers
        
        # Initialize and run EchoInStone
        app = EchoInStoneApp(output_dir=output_dir)
        result = app.process_audio_sync(echo_input, **diarization_options)
        
        # Add job ID to result
        result['job_id'] = job_id
        
        logger.info(f"Job {job_id} | Completed with status: {result.get('status', 'unknown')}")
        return result
        
    except Exception as e:
        error_msg = f'Processing failed: {str(e)}'
        logger.error(f"Job {job_id} | Handler error: {error_msg}")
        logger.exception("Full traceback:")
        
        return {
            'status': 'error',
            'error': error_msg,
            'job_id': job_id
        }

# Start the RunPod serverless worker
runpod.serverless.start({"handler": handler})

**Comment by johnx102** [comment]
Ok now , I didn't find where u can put the max, min speaker for diarization ...

**Comment by johnx102** [comment]
Hi @jeanjerome I'm almost done, I understand what you tell and finish my test and after I will send you all my modification.
Just one last question,i have an error when I try to switch whisper model from whisper-large-v3-turbo to whisper-large-v2. I have error can't load model for /openai.whisper-large-v2 ... but I have access to to this repisotery on hugging face. o do this modification on EchoInStone/processing/whisper_audio_transcriber.py, did you think I miss something? without this all is working great when I use whisper-large-v3-turbo

**Comment by jeanjerome** [comment]
Hi @johnx102 ,

To use `whisper large v2`, you have to instantiate:
 ```python
self._transcriber = WhisperAudioTranscriber(model_name="openai/whisper-large-v2")
``` 

(at https://github.com/jeanjerome/EchoInStone/blob/649079df3a2dc3e8c1d6cd8364fe6df199225f31/EchoInStone/app.py#L34C13-L34C58)
